Add new features to TinglingGit

adnrs96 opened the pull request:
In this PR, we add new options to this tooling. 
These options are
```
  --recursive           Using this we can walk the current Directory
                        recursively for checking files.
  --selective           Using this we can provide a selective list of files as
                        input which will be classified as safe and unsafe.
                        Generally this option is used with a file redirected
                        to input stream containing file names seperated by
                        newline char. Note:End of input is marked by inputing
                        a '$' symbol.
  --safe-rep            Using this we can perform repetitive tasks upon
                        different files and commit only changes which are safe
                        with respect to causing merge conflicts to other PR's.
                        The code to handle repetitive task could be placed by
                        developer under execute_task() function.
  --rebase-check        Using this while performing repetitive tasks, we can
                        do a rebase over PR's which might get conflicted by
                        changes to a particular file and see if that actually
                        happens and commit or revert the commit accordingly.
```

adnrs96 commented:
For doing the Text -> str migration I wrote the following script and inserted it in the designate place (as commented in code)
```python
    # Remove this return statement after adding custom SCRIPT

    COMMON_PREFIX_COMMIT_MSG = ': Change use of typing.Text to str.'
    # NOTE: Do not modify values of fn, associated_prs or root_dir in the custom script.
    # #######SCRIPT BEGINS HERE#########
    file_path, file_ext = os.path.splitext(fn)
    if file_ext != '.py' or 'migrations' in fn or fn.startswith('tools'):
        print('Skipping %s' % (fn))
        return -1

    import re
    import pyperclip
    fns = '/home/adnrs96/zulip/' + fn
    file = open(fns)
    original_file_input = file.read()
    text = original_file_input.split('\n')
    otext = text
    mtext = []
    pat = re.compile('Text')
    for ln in text:
      m = pat.sub('str', ln)
      mtext.append(m)
    text = mtext
    mtext = []
    pat1 = re.compile(', str')
    pat2 = re.compile('str, ')
    for ln in text:
      if 'from typing import' in ln and 'str' in ln:
          m = pat1.sub('', ln)
          m = pat2.sub('', m)
          if 'str' in m:
              m = 'SUPERGIRL'
          mtext.append(m)
          print(m)
      else:
          mtext.append(ln)
    text = '\n'.join(mtext)
    pat3 = re.compile('\nSUPERGIRL\n')
    text = pat3.sub('\n', text)
    pat4 = re.compile('SUPERGIRL\n')
    text = pat4.sub('\n', text)
    mtext = text.split('\n')


    file.close()

    final_file_output = '\n'.join(mtext)
    file = open(fns, 'w')
    file.write(final_file_output)
    file.close()
    print(fn)

    if original_file_input == final_file_output:
        return -1
    pyperclip.copy(fn)
    ####################################
    # #######SCRIPT ENDS HERE###########

    # raw_input() to wait for review upon changes and  proceed further
    xxx = raw_input()
    if xxx == '0':
        file = open(fns, 'w')
        file.write('\n'.join(otext))
        file.close()
        return -1

    # Commit changes done to file
    commit_cur_change = subprocess.Popen(
                        ['git', 'add', os.path.join(root_dir,fn)],
                         stdout=subprocess.PIPE,
                         stderr=subprocess.STDOUT).stdout.read().strip()
    print(commit_cur_change)
    commit_cur_change = subprocess.Popen(
                        ['git', 'commit', '-m', '/'.join(fn.split('/')[-2:]) + COMMON_PREFIX_COMMIT_MSG],
                         stdout=subprocess.PIPE,
                         stderr=subprocess.STDOUT).stdout.read().strip()
    commit_status = 1
```

timabbott commented on analyse-files.py:
```diff
@@ -10,6 +10,78 @@
+        print(t)
+        t = subprocess.Popen(('git checkout test-%s' % (prnum)).split(' '),
+                         stdout=subprocess.PIPE,
+                         stderr=subprocess.STDOUT).stdout.read().strip()
```
check out `subprocess.check_output` for this purpose.

timabbott commented:
I posted one comment, but otherwise this looks good.  Feel free to merge after cleaning that up.

adnrs96 commented:
@timabbott  Addressed your comment by modifying code in PR to use subprocess.check_output instead of subprocess.Popen and adding a separate commit to address other occurrences in code.

Merging now! Thanks for the review :)

timabbott commented:
Great, thanks @adnrs96!
